lib/ecs_deploy/auto_scaler.rb

```diff
@@ -21,70 +24,49 @@ def run(yaml_path, log_file = nil, error_log_file = nil)
+          polling_ths = @config["spot_instance_intrp_warns_queue_urls"].map do |queue_url|
+            Thread.new(queue_url) do |url|
+              drainer.poll_spot_instance_interruption_warnings(url)
+            end.tap { |th| th.abort_on_exception = true }
+          end
         end
 
         ths.each(&:join)
+
+        drainer&.stop
+        polling_ths&.each(&:join)
       end
 
-      def main_loop(asg_config, configs)
-        loop_with_polling_interval("loop of #{asg_config.name}") do
+      def main_loop(cluster_scaling_config, configs)
+        loop_with_polling_interval("loop of #{cluster_scaling_config.name}") do
           ths = configs.map do |service_config|
             Thread.new(service_config) do |s|
               @logger.debug "Start service scaling of #{s.name}"
-
-              if s.idle?
-                @logger.debug "#{s.name} is idling"
-                next
-              end
-
-              difference = 0
-              s.upscale_triggers.each do |trigger|
-                step = trigger.step || s.step
-                next if difference >= step
-
-                if trigger.match?
-                  logger.info "Fire upscale trigger of #{s.name} by #{trigger.alarm_name} #{trigger.state}"
-                  difference = step
-                end
-              end
-
-              if difference == 0 && s.desired_count > s.current_min_task_count
-                s.downscale_triggers.each do |trigger|
-                  next unless trigger.match?
-
-                  logger.info "Fire downscale trigger of #{s.name} by #{trigger.alarm_name} #{trigger.state}"
-                  step = trigger.step || s.step
-                  difference = [difference, -step].min
-                end
-              end
-
-              if s.current_min_task_count > s.desired_count + difference
-                difference = s.current_min_task_count - s.desired_count
-              end
-
-              if difference >= 0 && s.desired_count > s.max_task_count.max
-                difference = s.max_task_count.max - s.desired_count
-              end
-
-              if difference != 0
-                s.update_service(difference)
-              end
+              s.adjust_desired_count
             end
           end
           ths.each { |th| th.abort_on_exception = true }
 
           ths.each(&:join)
 
-          @logger.debug "Start asg scaling of #{asg_config.name}"
+          @logger.debug "Start cluster scaling of #{cluster_scaling_config.name}"
 
-          total_service_count = configs.inject(0) { |sum, s| sum + s.desired_count }
-          asg_config.update_auto_scaling_group(total_service_count, configs[0])
-          asg_config.detach_and_terminate_orphan_instances(configs[0])
+          required_capacity = configs.inject(0) { |sum, s| sum + s.desired_count * s.required_capacity }
+          cluster_scaling_config.update_desired_capacity(required_capacity, configs[0])
```

Comment:
I think that there is one probrem.

It seems to modify desired count of ECS service before autoscaler modifies desired capacity of Spot Fleet.
If autoscaler decreases count of ECS service, spot fleet instance is drained after a ECS task is terminated.
But we cannot control targets that are to be drained. In other words, instances that has a running task may be drained.
Because of this, Extra ECS services is terminated by instance draining.

What do you think about it?

Reply:
You're right, but I think it doesn't matter in practice for the following reasons:

* The cluster has enough resources to launch extra tasks when autoscaler decreases the desired tasks of the ECS service, so new tasks will be launched usually within 1 minute even if some container instances are drained.
* The service is likely to have a little extra tasks when downscale is triggered, so the service has enough tasks even if some tasks terminated by instance draining temporarily.

However, I have a concern that upscale might be triggered easily. I think users of autoscaler should adjust triggers in such a case.